src/KafkaNetClient/BrokerRouter.cs

```diff
@@ -173,6 +176,59 @@ public Task<bool> RefreshTopicMetadata(params string[] topics)
             return true;
         }
 
+        /// <summary>
+        /// Returns Topic metadata for each topic.
+        /// </summary>
+        /// <returns>List of Topics as provided by Kafka.</returns>
+        /// <remarks>
+        /// The topic metadata will by default check the cache. To force querying the metadata from the server use <see cref="RefreshAllTopicMetadata"/>
+        /// </remarks>
+        public List<Topic> GetAllTopicMetadataFromLocalCache()
+        {
+            return _topicIndex.Values.Select(t => t.Item1).ToList();
+        }
+
+        /// <summary>
+        /// Force a call to the kafka servers to refresh metadata for all topics.
+        /// </summary>
+        /// <remarks>
+        /// This method will ignore the cache and initiate a call to the kafka servers for all topics, updating the cache with the resulting metadata.
+        /// Only call this method to force a metadata update. For all other queries use <see cref="GetAllTopicMetadataFromLocalCache"/> which uses cached values.
+        /// </remarks>
+        public Task<bool> RefreshAllTopicMetadata()
+        {
+            return RefreshAllTopicMetadata(_kafkaOptions.RefreshMetadataTimeout);
+        }
+
+        /// <summary>
+        /// Force refresh each topic metadata that exists on kafka servers.
+        /// </summary>
+        private async Task<bool> RefreshAllTopicMetadata(TimeSpan timeout)
+        {
+            try
+            {
+                await _taskLocker.WaitAsync(timeout).ConfigureAwait(false);
+                _kafkaOptions.Log.DebugFormat("BrokerRouter: Refreshing metadata for all topics");
+
+                var connections = GetConnections();
+                var taskMetadata = _kafkaMetadataProvider.Get(connections);
+                await Task.WhenAny(Task.Delay(timeout), taskMetadata).ConfigureAwait(false);
```

Comment:
I would suggest to define and use extension method on `Task<T>`, `WithTimeout`, like here: https://github.com/dotnet/orleans/blob/master/src/Orleans/Async/TaskExtensions.cs#L218

It will shorten the code significantly.
You will be able to replace lines 215-222 with:
`var metadataResponse = await taskMetadata.WithTimeout(timeout).ConfigureAwait(false);`